Upgrade to RDSS MAS v4.0.0

sevein opened the pull request:
Updates according to the latest release of the spec:
https://github.com/JiscRDSS/rdss-message-api-specification/releases/tag/4.0.0

- Remove deprecated `Vocabulary` types, errors and service
- Message type and class now using generated enumerations
- Support new `PreservationEvent` message and `InformationPackage` model
- Update attribute names according to changes in API (pluralization, etc)
- Enumerations now generated via `enumeration.go`
- Remove unused type `message.MessageHandler`
- Add new `tenantJiscID` field to `MessageHeader`
- Add `ResearchObjectBase` to marshal/unmarshal research object subtypes, e.g. Article, Dataset or ThesisDissertation.
- Test decoding of new message samples provided in the spec repo
- Other changes, e.g. `dateValue` change to timestamp, new required fields, etc...

---

This PR is not using Go modules yet (the migration is happening in #111 instead).
Usage instructions before the migration:

#### Confirm that you're using Go 1.12

    $ go version
    go version go1.12.5 linux/amd64

#### Download the code

You can also use `git clone`.
But the location (`~/go/src/github.com/JiscRDSS/rdss-archivematica-channel-adapter`) is important when not using Go modules.

    $ env GO111MODULE=off go get github.com/JiscRDSS/rdss-archivematica-channel-adapter

#### Check out this branch

    $ cd ~/go/src/github.com/JiscRDSS/rdss-archivematica-channel-adapter
    $ git checkout dev/mas-update-v4.0.0

#### Clone the spec submodule

    $ git submodule update --init --recursive

#### Run tests and benchmarks

```
$ go test ./broker/message
ok  	github.com/JiscRDSS/rdss-archivematica-channel-adapter/broker/message	0.025s
```

```
$ go test -bench=. ./broker/message/
goos: linux
goarch: amd64
pkg: github.com/JiscRDSS/rdss-archivematica-channel-adapter/broker/message
BenchmarkValidationNoOp-6          	50000000	        47.0 ns/op
BenchmarkValidationSimple-6        	  300000	      6526 ns/op
BenchmarkValidationComplex-6       	  100000	     15005 ns/op
BenchmarkValidationVeryComplex-6   	   10000	    141058 ns/op
PASS
ok  	github.com/JiscRDSS/rdss-archivematica-channel-adapter/broker/message	7.552s
```

Note: the message-api-spec submodule is embedded into the binary. See `hack/build-spec.sh`.

sevein commented on broker/message/handler.go:
```diff
@@ -1,4 +0,0 @@
-package message
-
-// MessageHandler is a callback function that processes messages.
-type MessageHandler func(msg *Message)
```
Dead code.

sevein commented on broker/message/headers.go:
```diff
@@ -22,118 +16,7 @@ type MessageHeader struct {
-		return fmt.Errorf("invalid message type %q", s)
-	}
-	*t = v
-	return nil
```
`MessageType` and `MessageClass` removed, now using the generated enum types instead.

sevein commented on broker/message/headers.go:
```diff
@@ -22,118 +16,7 @@ type MessageHeader struct {
-	}
-	*t = v
-	return nil
+	TenantJiscID     uint64           `json:"tenantJiscID"`
```
This is what going to enable us to have a multi-tenant channel adapter!

sevein commented on broker/message/message.go:
```diff
@@ -1,5 +1,7 @@
 package message
 
+//go:generate go run generator.go
```
This is the magic behind the `go generate ./...` workflow.

cole commented on broker/message/doc.go:
```diff
@@ -0,0 +1,17 @@
+In order to update to the latest release:
+
+- Update the commit of the message-api-spec module so it points to the tag
+  desired.
```
This linebreaking looks a bit odd in godoc, it's wrapping desired in `<pre>` because of the indent.

sevein replied:
Thanks good catch. I've been reading https://blog.golang.org/godoc-documenting-go-code. Also this is useful https://github.com/fluhus/godoc-tricks. Changes: https://github.com/JiscRDSS/rdss-archivematica-channel-adapter/pull/110/commits/931de50b1c971201d8b1d8d6795b5363a42f1fc5.

cole commented on broker/message/message_preservation.go:
```diff
@@ -0,0 +1,21 @@
+	"fmt"
+)
+
+// Preservation Event
```
Should `PreservationEvent` be here instead of in `model_information_package`?

sevein replied:
I've been trying to keep message types and content model types separate since the latter belong to a different [repo](https://github.com/JiscRDSS/Canonical-data-model), just for clarity.

cole commented on broker/message/shared_intelectual_asset.go:
```diff
@@ -0,0 +1,57 @@
+	FileCompositionLevel   string              `json:"fileCompositionLevel"`
+	FileHasMimeType        bool                `json:"fileHasMimeType,omitempty"`
+	FileDateModified       []Timestamp         `json:"fileDateModified"`
+	FilePuid               []string            `json:"filePuid,omitempty"`
```
It seems odd that `FilePuid` isn't consistent with `FileUUID`, but I guess that matches the JSON?

sevein replied:
Fixed in 2c91da2.

sevein commented:
Thanks, @cole. I've addressed your comments. Also, published a new commit after updates in the spec repo: https://github.com/JiscRDSS/rdss-archivematica-channel-adapter/pull/110/commits/8a113345c4b33476deb7d7ed16f5cf12c6053bd9.
